kythe/go/platform/analysis/local/local.go

```diff
@@ -140,3 +144,98 @@ type kzipFetcher struct{ r *kzip.Reader }
 
 // Fetch implements the required method of analysis.Fetcher.
 func (k kzipFetcher) Fetch(_, digest string) ([]byte, error) { return k.r.ReadAll(digest) }
+
+// Analyzer provides an implementation of the analysis.CompilationAnalyzer
+// interface that works by saving the compilation unit to a temporary kzip file
+// and invoking an indexer binary on it.
+type Analyzer struct {
+	cmd     []string
+	fetcher analysis.Fetcher
+}
+
+// NewAnalyzer creates an analyzer that will execute the given command when
+// Analyze() is called. The filepath to a kzip is appended to the command, so if
+// given ["my_indexer", "--index"], the final executed command will be
+// ["my_indexer", "--index", "some/file.kzip"]. The indexer command should emit
+// length-delimited Entry protos to stdout as described in
+// //kythe/go/platform/delimited.
+func NewAnalyzer(cmd []string, fetcher analysis.Fetcher) *Analyzer {
+	if len(cmd) == 0 {
+		panic("parameter 'cmd' must not be empty")
+	}
+	return &Analyzer{cmd: cmd, fetcher: fetcher}
+}
+
+// Analyze implements the analysis.CompilationAnalyzer interface.
+func (a *Analyzer) Analyze(ctx context.Context, req *apb.AnalysisRequest, f analysis.OutputFunc) error {
+	tmpkzip, err := saveSingleUnitKzip(ctx, req.Compilation, a.fetcher)
+	if err != nil {
+		return fmt.Errorf("creating tmp kzip: %v", err)
+	}
+	defer os.Remove(tmpkzip)
+
+	// Run indexer binary
+	out, err := exec.Command(a.cmd[0], append(a.cmd[1:], tmpkzip)...).Output()
+	if e, ok := err.(*exec.ExitError); ok {
+		return fmt.Errorf("indexing kzip %s: %v (stderr=%s)", tmpkzip, err, string(e.Stderr))
+	} else if err != nil {
+		return fmt.Errorf("indexing kzip %s: %v", tmpkzip, err)
+	}
+
+	// Deserialize output and pass it to `f`.
+	rd := delimited.NewReader(bytes.NewReader(out))
+	for {
+		n, err := rd.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return fmt.Errorf("parsing entry stream: %v", err)
+		}
+
+		if err := f(ctx, &apb.AnalysisOutput{Value: n}); err != nil {
+			return fmt.Errorf("unable to write analysis output: %v", err)
+		}
+	}
+	if err := f(ctx, &apb.AnalysisOutput{
```

Comment:
Arguably this could as readily be:

```go
return f(ctx, &apb.AnalysisOutput{
   FinalResult: &apb.AnalysisResult{Status: apb.AnalysisResult_COMPLETE}.
})
```

The additional annotation probably doesn't add much value, since the caller already controls both the callback and the analyzer and therefore knows _why_ the error.

Reply:
:+1: